src/interface/MediaSource.js

```diff
@@ -0,0 +1,140 @@
+import PlaybackState from '../PlaybackState';
+
+import { debug } from '../utils/logger';
+import suppressAbortError from '../utils/suppressAbortError';
+
+// A reference to an HTML5 `<audio>` element. Currently this interface is a bit
+// fragile and invasive -- it assumes a single `<audio>` element, stores a
+// reference to it, and mutates it. If this element is managed by another
+// system simultaneously you may get strange behavior.
+//
+// TODO: Should the library inject its own `<audio>` element?
+let audioElement = null;
+
+// A reference to an HTML5 `MediaSource` instance.
+let source = null;
+
+// A reference to the internal MPEG buffer for the `source`. Appending MPEG data
+// to this buffer results in it being played.
+let sourceBuffer = null;
+
+// While playing, the system uses a recursive `setTimeout` loop to continuously
+// append audio data to the `sourceBuffer`. This is a reference to that timeout
+// and only exists while audio is being played.
+let sourceTimeout = null;
+
+function initialize({ volume = 1.0 }) {
+  audioElement = document.querySelector('audio');
+  source = new MediaSource();
+  sourceBuffer = null;
+  sourceTimeout = null;
+
+  setVolume(volume);
+}
+
+function play(clip) {
+  source = new MediaSource();
+  source.addEventListener('sourceopen', function () {
+    console.log('source open');
+    sourceBuffer = this.addSourceBuffer('audio/mpeg');
+    _play(clip);
+  });
+
+  audioElement.src = URL.createObjectURL(source);
+
+  return audioElement.play().catch(suppressAbortError);
+}
+
+function _play(clip) {
+  if (clip._clipState.chunksBufferingFinished) {
+    debug('MediaSourceEngine endOfStream()');
+    source.endOfStream();
+    return;
+  }
+
+  const isChunkReady = clip._clipState.isChunkReady(clip._clipState.chunkIndex);
+
+  const useSilence =
+    !isChunkReady &&
+    clip._clipState.chunkIndex === clip._initialChunk &&
+    !clip._wasPlayingSilence &&
+    (clip.browserName === 'safari' || clip.osName === 'ios');
```

Comment:
Maybe we could have a new method on `Clip` called `shouldUseSilence` or something? Also, not sure if we should have the `browserName` and `osName` on the `Clip`, it'd be nice if those values are defined on the `ProtonPlayer` level